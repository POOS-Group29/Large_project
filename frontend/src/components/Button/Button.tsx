import clsx from "clsx";
import React from "react";

import { Spinner } from "../Spinner";

const variants = {
  primary:
    "bg-blue-600 text-white hover:bg-blue-500  focus-visible:outline-blue-600",
  inverse:
    "bg-white text-black border border-gray-300 hover:bg-gray-100 focus-visible:outline-gray-300",
  danger:
    "bg-red-600 text-white hover:bg-red-500 focus-visible:outline-red-600",
};

const sizes = {
  sm: "px-3 py-1.5 text-sm",
  md: "px-6 py-2 text-md",
  lg: "px-8 py-3 text-lg",
};

type IconProps =
  | { startIcon: React.ReactElement; endIcon?: never }
  | { endIcon: React.ReactElement; startIcon?: never }
  | { endIcon?: undefined; startIcon?: undefined };

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: keyof typeof variants;
  size?: keyof typeof sizes;
  isLoading?: boolean;
} & IconProps;

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      type = "button",
      className = "",
      variant = "primary",
      size = "md",
      isLoading = false,
      startIcon,
      endIcon,
      ...props
    },
    ref
  ) => {
    return (
      <button
        ref={ref}
        type={type}
        className={clsx(
          "flex justify-center items-center font-semibold leading-6 border border-gray-300 disabled:opacity-70 disabled:cursor-not-allowed rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2",
          variants[variant],
          sizes[size],
          className
        )}
        {...props}
      >
        {isLoading && <Spinner size="sm" className="text-current" />}
        {!isLoading && startIcon}
        <span className="mx-2" data-testid="button-text">
          {props.children}
        </span>{" "}
        {!isLoading && endIcon}
      </button>
    );
  }
);

Button.displayName = "Button";
