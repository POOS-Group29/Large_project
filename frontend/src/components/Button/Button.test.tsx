import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { Button } from "./Button";

describe("Button", () => {
  test("should match snapshot", () => {
    const { container } = render(<Button />);
    expect(container).toMatchSnapshot();
  });

  test("should have spinner and text if isLoading is true", () => {
    const text = "Click me";
    const { container } = render(<Button isLoading>{text}</Button>);
    expect(container).toMatchSnapshot();
    expect(
      container.querySelector("[data-testid='spinner']")
    ).toBeInTheDocument();
    expect(
      container.querySelector("[data-testid='button-text']")
    ).toHaveTextContent(text);
  });

  test("should have text if isLoading is false", () => {
    const text = "Click me";
    const { container } = render(<Button>{text}</Button>);
    expect(container).toMatchSnapshot();
    expect(
      container.querySelector("[data-testid='spinner']")
    ).not.toBeInTheDocument();
    expect(
      container.querySelector("[data-testid='button-text']")
    ).toHaveTextContent(text);
  });
});
