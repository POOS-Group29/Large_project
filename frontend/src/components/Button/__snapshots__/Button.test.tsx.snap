// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should have spinner and text if isLoading is true 1`] = `
<div>
  <button
    class="flex justify-center items-center font-semibold leading-6 border border-gray-300 disabled:opacity-70 disabled:cursor-not-allowed rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-blue-600 text-white hover:bg-blue-500  focus-visible:outline-blue-600 px-6 py-2 text-md"
    type="button"
  >
    <svg
      class="animate-spin h-4 w-4 text-blue-600 text-current"
      data-testid="spinner"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
    <span
      class="sr-only"
    >
      Loading
    </span>
    <span
      class="mx-2"
      data-testid="button-text"
    >
      Click me
    </span>
     
  </button>
</div>
`;

exports[`Button should have text if isLoading is false 1`] = `
<div>
  <button
    class="flex justify-center items-center font-semibold leading-6 border border-gray-300 disabled:opacity-70 disabled:cursor-not-allowed rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-blue-600 text-white hover:bg-blue-500  focus-visible:outline-blue-600 px-6 py-2 text-md"
    type="button"
  >
    <span
      class="mx-2"
      data-testid="button-text"
    >
      Click me
    </span>
     
  </button>
</div>
`;

exports[`Button should match snapshot 1`] = `
<div>
  <button
    class="flex justify-center items-center font-semibold leading-6 border border-gray-300 disabled:opacity-70 disabled:cursor-not-allowed rounded-md shadow-sm focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 bg-blue-600 text-white hover:bg-blue-500  focus-visible:outline-blue-600 px-6 py-2 text-md"
    type="button"
  >
    <span
      class="mx-2"
      data-testid="button-text"
    />
     
  </button>
</div>
`;
